### ISSUE TITLE
[pub_resolver] misleading error running `pub get`

### ISSUE THREAD

**Issue by iapicca** [dart-lang/pub#3494]
### Environment

<details>
  <summary>doctor</summary>
 
```bash
[✓] Flutter (Channel stable, 3.0.5, on macOS 12.4 21F79 darwin-arm, locale en-EE)
[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
[✓] Xcode - develop for iOS and macOS (Xcode 13.4.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2021.2)
[✓] VS Code (version 1.69.1)
[✓] Connected device (2 available)
[✓] HTTP Host Availability

• No issues found!
```

</details>

## Steps to Reproduce

1. create a **flutter package**

2.  add to pubspec any package that depend from 
```yaml
  meta: ^1.8.0
```

3. run `flutter pub get` and observe the error
```console
Running "flutter pub get" in <APP>...                       
Because every version of flutter_test from sdk depends on meta 1.7.0 and <PACKAGE> >=0.1.0+4 depends on meta ^1.8.0, flutter_test from sdk is incompatible with <PACKAGE> >=0.1.0+4.
So, because <APP> depends on both <PACKAGE> ^1.0.0 and flutter_test from sdk, version solving failed.
pub get failed (1; So, because <APP> depends on both <PACKAGE> ^1.0.0 and flutter_test from sdk, version solving failed.)
exit code 1
```

4. add `dependency overrides`
```yaml
 dependency_overrides:
   meta: ^<some version>
```

5. run `dart pub publish --dry-run` and observe the error below,
please note the [misleading error message](https://github.com/dart-lang/pub/issues/3493#issue-1306010197), as `flutter_test` is **not** a non-dev dependency
```console
Package validation found the following error:
* Your pubspec.yaml must not override non-dev dependencies.
  This ensures you test your package against the same versions of its dependencies
  that users will have when they use it.
Sorry, your package is missing a requirement and can't be published yet.
For more information, see: https://dart.dev/tools/pub/cmd/pub-lish.
```

6. remove `flutter_test` dependency, run once more`flutter pub get` and observe the error 
```console
Resolving dependencies in ... (1.7s)
Because <PACKAGE> >=0.1.0+4 depends on meta ^1.8.0 and every version of flutter from sdk depends on meta 1.7.0, <PACKAGE> >=0.1.0+4 is incompatible with flutter from sdk.
So, because yak_tween depends on both flutter from sdk and <PACKAGE> ^1.0.0, version solving failed.
```

## Expected behavior

`pub get` command should provide the correct error to the user (as in _**steps to reproduce - 6**_)

## Actual behavior

`pub get` command provided misleading error

#

see [answer](https://github.com/flutter/flutter/issues/107717#issuecomment-1185728964)

**Comment by sigurdm** [comment]
Sorry - I find this a bit hard to follow. Which message is misleading? The output from `pub get` (that seems OK to me).

> please note the https://github.com/dart-lang/pub/issues/3493#issue-1306010197, as flutter_test is not a non-dev dependency

`package:meta` is not a dev-dependency (it is a transitive dependency), and therefore we don't allow overriding it when publishing.
